File: prototype/Assets/microcosmicWar/Scripts/zz/zzItemBagControl.js

var bagIndex=0;
//若为真 代表是已有的包,未设置就用bagName创建新包
var useExistBag = false;
var bagName="";
var owner:GameObject;

protected var itemSystem:zzItemSystem;
protected var haveBeInited = false;

//为了解决初次刷新时 UI和此类 初始化先后的问题
var callListAfterStart = Array();

function setUseExistBag(pBagIndex:int)
{
	//useExistBag = pUse;
	//bagIndex = pBagIndex;
	
	zzCreatorUtility.sendMessage(gameObject,"impSetUseExistBag",pBagIndex);
}

@RPC
function impSetUseExistBag(pBagIndex:int)
{
	useExistBag = true;
	bagIndex = pBagIndex;
}

function addCallAfterStart(pCall)
{
	if(haveBeInited)
		pCall();
	else
		callListAfterStart.Add(pCall);
}

protected function callAfterStart()
{
	for(var call in callListAfterStart)
		call();
	callListAfterStart.Clear();
}


function getItemSystem()
{
	return itemSystem;
}

class zzMyItemInBagInfo
{
	var name:String;
	var number:int;
}

var itemInBagInfo:zzMyItemInBagInfo[];

function setItemChangedCall(pCall)
{
	getBagData().addDataChangedCall(pCall,1,getBagData().getItemTypeNum()-1);
}

function setMoneyChangedCall(pCall)
{
	getBagData().addDataChangedCall(pCall,0,0);
}

function getBagData():ItemBagData
{
	return  itemSystem.getBagTable().getData(bagIndex);
}

function setItemNum(pName:String,pNum:int)
{
	getBagData().setNum(itemSystem.getItemTypeTable().getIndex(pName),pNum);
}

//0 为钱
function addItemOne(pItemIndex:int)
{
	getBagData().addItemOne(pItemIndex);
}

//0 为钱
function addItem(index:int,number:int)
{
	getBagData().addItem(index,number);
}

//得到除索引为0以外的工具数组,以便UI显示
function getItemList()
{
	var lOut = Array();
	for(var i=1;i<itemSystem.getItemTypeTable().getNum();++i)
	{
		var numOfTheItem = getNum(i);
		while(numOfTheItem>0)
		{
			lOut.Add(i);
			--numOfTheItem;
		}
	}
	return lOut;
}

function addMoney(number:int)
{
	//Debug.Log("addMoney");
	addItem(0,number);
	//Debug.Log(getMoneyNum());
}

function getNum(index:int)
{
	return getBagData().getNum(index);
}

function getMoneyNum()
{
	return getNum(0);
}

function useItemOne(pItemIndex:int,pOwner:GameObject)
{
	//print("index:"+pItemIndex+",number:"+getBagData().getNum(pItemIndex));
	if( zzCreatorUtility.isHost() )
		impUseItemOne(pItemIndex,pOwner);
	else
		networkView.RPC( "RPCUseItemOne", RPCMode.Server ,pItemIndex , networkView.viewID );
}

@RPC
function RPCUseItemOne(pItemIndex:int,pOwnerID:NetworkViewID)
{
	impUseItemOne(pItemIndex,NetworkView.Find(pOwnerID).gameObject);
}

protected function impUseItemOne(pItemIndex:int,pOwner:GameObject)
{
	if(getNum(pItemIndex)>0)
	{
		var item:ItemObjectImp = itemSystem.getItemTypeTable().getData(pItemIndex).getItemObject();
		if(item.canUse(pOwner))
		{
			item.use();
			getBagData().removeItemOne(pItemIndex);
		}
	}
}

function useItemOne(pItemIndex:int)
{
	useItemOne(pItemIndex,owner);
}

function getBagID()
{
	return bagIndex;
}

function Start()
{
	if(!owner)
		owner= gameObject;
	itemSystem = zzItemSystem.getSingleton();
	if(useExistBag)
	{
		//print("bagIndex!=-1");
		//print(bagIndex);
		bagName = getBagData().name;
	}
	else
	{
		bagIndex =itemSystem.addBag(bagName);
		//print(bagIndex);
		for(var i:zzMyItemInBagInfo in itemInBagInfo)
		{
			setItemNum(i.name,i.number);
		}
	}
		
	
	//和发射器连接 以便得到奖励
	var lEmitter:Emitter = owner.GetComponentInChildren(Emitter);
	if(lEmitter)
		lEmitter.setInjureInfo({"bagControl":this});
	
	//Debug.Log(getBagData());
	
	haveBeInited = true;
	callAfterStart();
}

File: prototype/Assets/microcosmicWar/Scripts/zz/zzItemBagControl.cs

using UnityEngine;
using System.Collections;

public class zzItemBagControl : MonoBehaviour
{


    public int bagIndex = 0;
    //若为真 代表是已有的包,未设置就用bagName创建新包
    public bool useExistBag = false;
    public string bagName = "";
    public GameObject owner;

    protected zzItemSystem itemSystem;
    protected bool haveBeInited = false;

    //为了解决初次刷新时 UI和此类 初始化先后的问题
    public ArrayList callListAfterStart = new ArrayList();

    public void setUseExistBag(int pBagIndex)
    {
        //useExistBag = pUse;
        //bagIndex = pBagIndex;

        zzCreatorUtility.sendMessage(gameObject, "impSetUseExistBag", pBagIndex);
    }

    [RPC]
    public void impSetUseExistBag(int pBagIndex)
    {
        useExistBag = true;
        bagIndex = pBagIndex;
    }

    public void addCallAfterStart(zzUtilities.voidFunction pCall)
    {
        //for 转换
        zzUtilities.voidFunction lTempCall = pCall;
        lTempCall = pCall;
        if (haveBeInited)
            lTempCall();
        //pCall();
        else
            callListAfterStart.Add(pCall);
    }

    protected void callAfterStart()
    {
        //for 转换
        //zzUtilities.voidFunction lTempCall ;

        foreach (zzUtilities.voidFunction call in callListAfterStart)
        {
           // lTempCall = call;
            //lTempCall();
            call();
        }
        callListAfterStart.Clear();
    }


    public zzItemSystem getItemSystem()
    {
        return itemSystem;
    }

    [System.Serializable]
    public class zzMyItemInBagInfo
    {
        public string name;
        public int number;
    }

    public zzMyItemInBagInfo[] itemInBagInfo;

    public void setItemChangedCall(zzUtilities.voidFunction pCall)
    {
        getBagData().addDataChangedCall(pCall, 1, getBagData().getItemTypeNum() - 1);
    }

    public void setMoneyChangedCall(zzUtilities.voidFunction pCall)
    {
        getBagData().addDataChangedCall(pCall, 0, 0);
    }

    public ItemBagData getBagData()
    {
        return itemSystem.getBagTable().getData(bagIndex) as ItemBagData;
    }

    public void setItemNum(string pName, int pNum)
    {
        getBagData().setNum(itemSystem.getItemTypeTable().getIndex(pName), pNum);
    }

    //0 为钱
    public void addItemOne(int pItemIndex)
    {
        getBagData().addItemOne(pItemIndex);
    }

    //0 为钱
    public void addItem(int index, int number)
    {
        getBagData().addItem(index, number);
    }

    //得到除索引为0以外的工具数组,以便UI显示
    public ArrayList getItemList()
    {
        ArrayList lOut = new ArrayList();
        for (int i = 1; i < itemSystem.getItemTypeTable().getNum(); ++i)
        {
            int numOfTheItem = getNum(i);
            while (numOfTheItem > 0)
            {
                lOut.Add(i);
                --numOfTheItem;
            }
        }
        return lOut;
    }

    public void addMoney(int number)
    {
        //Debug.Log("addMoney");
        addItem(0, number);
        //Debug.Log(getMoneyNum());
    }

    public int getNum(int index)
    {
        return getBagData().getNum(index);
    }

    public int getMoneyNum()
    {
        return getNum(0);
    }

    public void useItemOne(int pItemIndex, GameObject pOwner)
    {
        //print("index:"+pItemIndex+",number:"+getBagData().getNum(pItemIndex));
        if (zzCreatorUtility.isHost())
            impUseItemOne(pItemIndex, pOwner);
        else
            networkView.RPC("RPCUseItemOne", RPCMode.Server, pItemIndex, networkView.viewID);
    }

    [RPC]
    public void RPCUseItemOne(int pItemIndex, NetworkViewID pOwnerID)
    {
        impUseItemOne(pItemIndex, NetworkView.Find(pOwnerID).gameObject);
    }

    protected void impUseItemOne(int pItemIndex, GameObject pOwner)
    {
        if (getNum(pItemIndex) > 0)
        {
            ItemTypeInfo lItemTypeInfo = itemSystem.getItemTypeTable()
                .getData(pItemIndex) as ItemTypeInfo;
            IitemObject item = lItemTypeInfo.getItemObject();
            if (item.canUse(pOwner))
            {
                item.use();
                getBagData().removeItemOne(pItemIndex);
            }
        }
    }

    public void useItemOne(int pItemIndex)
    {
        useItemOne(pItemIndex, owner);
    }

    public int getBagID()
    {
        return bagIndex;
    }

    void Start()
    {
        if (!owner)
            owner = gameObject;
        itemSystem = zzItemSystem.getSingleton();
        if (useExistBag)
        {
            //print("bagIndex!=-1");
            //print(bagIndex);
            bagName = getBagData().name;
        }
        else
        {
            bagIndex = itemSystem.addBag(bagName);
            //print(bagIndex);
            foreach (zzMyItemInBagInfo i in itemInBagInfo)
            {
                setItemNum(i.name, i.number);
            }
        }


        //和发射器连接 以便得到奖励
        Emitter lEmitter = owner.GetComponentInChildren<Emitter>();
        if (lEmitter)
        {
            Hashtable lInjureInfo = new Hashtable();
            lInjureInfo["bagControl"] = this;

            lEmitter.setInjureInfo(lInjureInfo);
            //lEmitter.setInjureInfo({"bagControl":this});
        }

        //Debug.Log(getBagData());

        haveBeInited = true;
        callAfterStart();
    }


}